### PR TITLE
Add support for Dash Masternode methods

### PR DESCRIPTION
**Pull request by thelazier**
This will add Dash specific feature which require #182 to be merged so it's also include #182 .

### CODE REVIEW

File: lib/coins.py [hunk 1]
```diff
@@ -594,6 +596,10 @@ class Dash(Coin):
         'electrum.dash.siampm.com s t',
         'wl4sfwq2hwxnodof.onion s t',
     ]
+    from server.session import DashElectrumX
+    SESSIONCLS = DashElectrumX
+    from server.daemon import DashDaemon
+    DAEMON = DashDaemon
```
Comment (kyuupichan) [inline lib/coins.py]: Just put these imports at the top please

File: server/session.py [hunk 2]
```diff
@@ -387,3 +387,69 @@ def __init__(self, *args, **kwargs):
...
+                    'method': 'masternode.subscribe',
+                    'params': [masternode],
+                    'result': status.get(masternode),
+                }
+                self.send_binary(self.encode_payload(payload))
+
+    def server_version(self, client_name=None, protocol_version=None):
+        '''Returns the server version as a string.
+        ForcE version string response for Electrum-Dash 2.6.4
+        '''
```
Comment (kyuupichan) [inline server/session.py]: ForcE typo.  Can you also please add a comment explaining why the special case is needed.

File: server/session.py [hunk 3]
```diff
@@ -387,3 +387,69 @@ def __init__(self, *args, **kwargs):
...
+    # Masternode command handlers
+    async def masternode_announce_broadcast(self, signmnb):
+        '''Pass through the parameters to the daemon.
+
+        An ugly API: current Electrum clients only pass the masternode
+        broadcast message in hex and expect error messages to be returned in
+        the result field.  And the server shouldn't be doing the client's
+        user interface job here.
+        '''
+        try:
```
Comment (kyuupichan) [inline server/session.py]: Is there a plan to fix this API?

File: server/session.py [hunk 4]
```diff
@@ -387,3 +387,69 @@ def __init__(self, *args, **kwargs):
...
+
+
+class DashElectrumX(ElectrumX):
+    '''A TCP server that handles incoming Electrum Dash connections.'''
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.electrumx_handlers['masternode.announce.broadcast'] = self.masternode_announce_broadcast
+        self.electrumx_handlers['masternode.subscribe'] = self.masternode_subscribe
+        self.subscribe_mns = False
```
Comment (kyuupichan) [inline server/session.py]: I don't think we need this boolean variable.  You can just loop over the mns set.

File: server/session.py [hunk 5]
```diff
@@ -387,3 +387,69 @@ def __init__(self, *args, **kwargs):
...
+        self.mns = set()
+
+    async def notify(self, height, touched):
+        '''Notify the client about changes in masternode list.'''
+
+        await super().notify(height, touched)
+
+        if self.subscribe_mns:
+            for masternode in self.mns:
```
Comment (kyuupichan) [inline server/session.py]: The condition should not be necessary right?

File: server/session.py [hunk 6]
```diff
@@ -387,3 +387,69 @@ def __init__(self, *args, **kwargs):
...
+            self.log_info('masternode_broadcast: {}'.format(message))
+            return (
+                'The masternode broadcast was rejected.  ({})\n[{}]'
+                .format(message, signmnb)
+            )
+
+    async def masternode_subscribe(self, vin):
+        '''Returns the status of masternode.'''
+        result = await self.daemon.masternode_list(['status',vin])
+        if result is not None:
```
Comment (kyuupichan) [inline server/session.py]: Space after comma

File: server/session.py [hunk 7]
```diff
@@ -387,3 +387,69 @@ def __init__(self, *args, **kwargs):
...
+            self.mns.add(vin)
+            self.subscribe_mns = True
+        return result.get(vin)
```
Comment (kyuupichan) [inline server/session.py]: If result can be None we cannot "get" it.  If it cannot be None we don't need to test.

File: lib/coins.py (hunk 1 above)
Comment (thelazier) [inline lib/coins.py]: Top Lines inside `class Dash(Coin):` , right?

File: server/session.py (hunk 3 above)
Comment (thelazier) [inline server/session.py]: I think this API won't be fixed, it should be similar to broadcast transaction and the daemon would do all the verified by network anyway. So, Comment is removed.

File: lib/coins.py (hunk 1 above)
Comment (thelazier) [inline lib/coins.py]: Done via 	8d21eae

File: server/session.py (hunk 4 above)
Comment (thelazier) [inline server/session.py]: Agree.

File: server/session.py (hunk 5 above)
Comment (thelazier) [inline server/session.py]: Agree.

File: server/session.py (hunk 2 above)
Comment (thelazier) [inline server/session.py]: Typo is fixed.

File: server/session.py (hunk 6 above)
Comment (thelazier) [inline server/session.py]: Yes, sir.

File: server/session.py (hunk 7 above)
Comment (thelazier) [inline server/session.py]: You are correct.

File: server/session.py [hunk 8]
```diff
@@ -387,3 +387,62 @@ def __init__(self, *args, **kwargs):
...
+                'method': 'masternode.subscribe',
+                'params': [masternode],
+                'result': status.get(masternode),
+            }
+            self.send_binary(self.encode_payload(payload))
+
+    def server_version(self, client_name=None, protocol_version=None):
+        '''Returns the server version as a string.
+        Force version string response for Electrum-Dash 2.6.4
+        '''
```
Comment (kyuupichan) [inline server/session.py]: Please comment on why the special case

Comment (thelazier) [inline server/session.py]: This is directly related to https://github.com/dashpay/electrum-dash/commit/638cf6c0aeb7be14a85ad98f873791cb7b49ee29 so this special case will be removed later when newer version of electrum-dash is released without this server version checking.

**Comment by kyuupichan** [comment]
I think the only thing missing is the move of the imports to the top of the file (not function) that I requested a couple of times.

**Comment by thelazier** [comment]
Moved and Pushed as https://github.com/thelazier/electrumx/commit/50c0cf66b1c897532c6002c92bc17fde7a030805

**Comment by kyuupichan** [comment]
Any reason you've not repushed as this PR so I can just click Merge?

**Comment by thelazier** [comment]
I pushed in same branch with this PR but I don't know why the PR is not update

**Comment by kyuupichan** [comment]
OK please fix that in any follow-ups you might have.

**Pull request merged**